src/implicit_function.jl

```diff
@@ -1,46 +1,134 @@
+"""
+struct Conditions{handle_byproduct,C}
+    c::C
+    function Conditions{handle_byproduct}(c::C) where {handle_byproduct,C}
+        return new{handle_byproduct,C}(c)
+    end
 end
 
+function Base.show(
+    io::IO, conditions::Conditions{handle_byproduct}
+) where {handle_byproduct}
+    return print("Conditions{$handle_byproduct}($(conditions.c))")
+end
+
+"""
+    conditions(x, y, z; kwargs...)
+
+Apply `conditions.c` to `(x, y, z)`, discarding `z` beforehand if needed.
 """
-    ImplicitFunction(forward, conditions)
+(conditions::Conditions{true})(x, y, z; kwargs...) = conditions.c(x, y, z; kwargs...)
+(conditions::Conditions{false})(x, y, z; kwargs...) = conditions.c(x, y; kwargs...)
 
-Construct an `ImplicitFunction` with `Krylov.gmres` as the default linear solver.
 """
-function ImplicitFunction(forward, conditions)
-    return ImplicitFunction(forward, conditions, gmres)
+    ImplicitFunction{handle_byproduct,FF<:Forward,CC<:Conditions,LS}
+
+Differentiable wrapper for an implicit function defined by a forward mapping and a set of conditions.
+
+# Details
+
+- If `handle_byproduct=false`, the forward mapping is `x -> y(x)` and the conditions are `c(x,y(x)) = 0`.
+- If `handle_byproduct=true`, the forward mapping is `x -> (y(x),z(x))` and the conditions are `c(x,y(x),z(x)) = 0`. In this case, `z(x)` contains additional information generated by the forward mapping, but considered constant for differentiation purposes.
+
+Given `x ∈ ℝⁿ` and `y ∈ ℝᵈ`, we need as many conditions as output dimensions: `c(x,y,z) ∈ ℝᵈ`. We can then compute the Jacobian of `y(⋅)` using the implicit function theorem:
+```
+∂₂c(x,y(x),z(x)) * ∂y(x) = -∂₁c(x,y(x),z(x))
+```
+This requires solving a linear system `A * J = -B`, where `A ∈ ℝᵈˣᵈ`, `B ∈ ℝᵈˣⁿ` and `J ∈ ℝᵈˣⁿ`.
+
+# Fields
+
+- `forward::FF`: a wrapper of type [`Forward`](@ref) coherent with the value of `handle_byproduct`
+- `conditions::FF`: a wrapper of type [`Conditions`](@ref) coherent with the value of `handle_byproduct`
+- `linear_solver::LS`: a callable of the form `(A,b) -> (u,stats)` such that `Au = b` and `stats.solved ∈ {true,false}`, typically taken from [Krylov.jl](https://github.com/JuliaSmoothOptimizers/Krylov.jl)
+
+# Constructors
+
+    ImplicitFunction(f, c; linear_solver=gmres)
+    ImplicitFunction(f, c, Val(handle_byproduct); linear_solver=gmres)
+
+Construct an `ImplicitFunction` by wrapping a forward mapping `f` into a field of type `Forward` and conditions `c` into a field of type `Conditions`, taking into account the value of `handle_byproduct` (which defaults to `false`).
+The default linear solver is `Krylov.gmres`, but this can be changed with a keyword argument.
+"""
+struct ImplicitFunction{
+    handle_byproduct,FF<:Forward{handle_byproduct},CC<:Conditions{handle_byproduct},LS
+}
+    forward::FF
+    conditions::CC
+    linear_solver::LS
+
+    function ImplicitFunction(
```

Comment:
This is probably better as an inner constructor since it enforces the invariant we want. That way users cannot screw up the `handle_byproduct` shenanigans